File: TodoTest.Web/Scripts/app/controllers/TodoController.js
Todo.controller('TodoController', function ($scope, $log, TodoService) {

    $scope.currentPage = 1;
    $scope.pageSize = 10;
    $scope.todos = [];
    $scope.newTodoName = "";

    $scope.init = function () {
        var promise = TodoService.get($scope.currentPage, $scope.pageSize);
        promise.then(function (response) {
            $scope.todos = response;
        });
    };

    $scope.addTodo = function () {
        if (!$scope.newTodoName) return;

        var promise = TodoService.add($scope.newTodoName);
        promise.then(function (response) {
            $scope.todos.push(new TodoViewModel(response.Id, $scope.newTodoName, false));
            $scope.newTodoName = "";
        });
    };

    $scope.removeTodo = function (id) {
        var promise = TodoService.remove(id);
        promise.then(function (response) {
            $scope.todos = _.reject($scope.todos, function (x) { return x.Id === id; });
        });
    };

    $scope.completeTodo = function(todo) {
        var promise = TodoService.update(todo);
        promise.then(function (response) {
            // ui updated by bindings
        });
    };

    $scope.hasAnyTodos = function () {
        return _.size($scope.todos) > 0;
    };

    function TodoViewModel(id, name, isComplete) {
        this.Id = id;
        this.Name = name;
        this.Completed = isComplete;
    }
});

File: TodoTest.Web/Controllers/TodoController.cs
using System;
using System.Linq;
using System.Net;
using System.Web;
using System.Web.Mvc;
using TodoTest.Web.Data;
using TodoTest.Web.Models;
using TodoTest.Web.Models.ViewModels;

namespace TodoTest.Web.Controllers
{
    public class TodoController : Controller
    {
        private readonly IStore store;

        public TodoController(IStore store)
        {
            this.store = store;
        }

        public JsonResult List(int? page, int? pageSize)
        {
            var todos = store.Get<Todo>(page.GetPage(), pageSize.GetPageSize());
            return new JsonResult
            {
                Data = todos.Select(x => new TodoViewModel(x)),
                JsonRequestBehavior = JsonRequestBehavior.AllowGet
            };
        }

        [HttpPost]
        public JsonResult Add(string name)
        {
            if (string.IsNullOrEmpty(name))
            {
                throw new HttpException((int)HttpStatusCode.BadRequest, "name cant be empty");
            }

            var createdTodo = store.Add(new Todo(name));
            return new JsonResult
            {
                Data = createdTodo
            };
        }

        [HttpPost]
        public JsonResult Remove(string guid)
        {
            var deletedToDo = store.Delete<Todo>(Guid.Parse(guid));
            if (deletedToDo == null)
            {
                throw new HttpException((int)HttpStatusCode.NotFound, "todo not found");
            }

            return new JsonResult
            {
                Data = new TodoViewModel(deletedToDo)
            };
        }

        [HttpPost]
        public JsonResult Update(TodoViewModel todo)
        {
            var updatedTodo = store.Update(new Todo(todo.Name, todo.Completed) { Id = Guid.Parse(todo.Id) });
            if (updatedTodo == null)
            {
                throw new HttpException((int)HttpStatusCode.NotFound, "todo not found");
            }

            return new JsonResult
            {
                Data = new TodoViewModel(updatedTodo)
            };
        }
    }
}
